src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketBranchClientImpl.java

```diff
@@ -36,34 +41,53 @@ public Stream<BitbucketDefaultBranch> getRemoteBranches() {
                 .addPathSegment("repos")
                 .addPathSegment(repositorySlug)
                 .addPathSegment("branches")
-                .addQueryParameter("limit", String.valueOf(PAGE_SIZE));
+                .addQueryParameter("limit", String.valueOf(PAGE_SIZE))
+                .addQueryParameter("orderBy", "modification"); // the most recently-modified branches first
 
         HttpUrl url = urlBuilder.build();
         BitbucketPage<BitbucketDefaultBranch> firstPage =
-                bitbucketRequestExecutor.makeGetRequest(url, new TypeReference<BitbucketPage<BitbucketDefaultBranch>>() {}).getBody();
-        return BitbucketPageStreamUtil.toStream(firstPage, new NextPageFetcherImpl(url, bitbucketRequestExecutor))
-                .map(BitbucketPage::getValues).flatMap(Collection::stream);
+                bitbucketRequestExecutor.makeGetRequest(url,
+                        new TypeReference<BitbucketPage<BitbucketDefaultBranch>>() {
+                }).getBody();
+        return BitbucketPageStreamUtil.toStream(firstPage, new NextPageFetcherImpl(url, bitbucketRequestExecutor,
+                        MAX_PAGES))
+                .map(BitbucketPage::getValues)
+                .flatMap(Collection::stream);
     }
 
     static class NextPageFetcherImpl implements NextPageFetcher<BitbucketDefaultBranch> {
 
         private final BitbucketRequestExecutor bitbucketRequestExecutor;
+        private final AtomicInteger currentPage = new AtomicInteger();
+        private final int maxPages;
         private final HttpUrl url;
 
         NextPageFetcherImpl(HttpUrl url,
-                            BitbucketRequestExecutor bitbucketRequestExecutor) {
+                            BitbucketRequestExecutor bitbucketRequestExecutor,
+                            int maxPages) {
             this.url = url;
             this.bitbucketRequestExecutor = bitbucketRequestExecutor;
+            this.maxPages = maxPages;
         }
 
         @Override
         public BitbucketPage<BitbucketDefaultBranch> next(BitbucketPage<BitbucketDefaultBranch> previous) {
             if (previous.isLastPage()) {
                 throw new IllegalArgumentException("Last page does not have next page");
             }
+
+            if (currentPage.incrementAndGet() >= maxPages) {
+                // We've reached the maximum number of pages, so we return an "empty last page" to stop the iterator.
```

Comment:
Add logging to inform users that the limit has been met.

Reply:
Depending on how you feel about it, you could modify the `BitbucketPage` to contain the information needed for logging. (I don't know if this is a good idea, maybe it is just an ugly hack)